### DISCUSSION TITLE
Dumb question about inserts and keeping the front end updated

### DISCUSSION THREAD

**Discussion by matthewlyf** [supabase/supabase#26536] [Questions] [answered]
I'm new to web app development so sorry in advance if I am unable to describe things properly, I'm still at the stage of "I don't know what I don't know". I have been stumped by a problem that I can't seem to understand how to resolve. I am testing a python script in collab (separate from my web app) to generate row data that I insert into several tables in Supabase. My webapp should then be reading from these tables, for example I have a metric that presents the number of rows from one of my tables or you can query values from a specific column to return values from the row. I've verified that the new rows are found in Supabase but oddly they aren't reflected in my front end. Is there a step that I'm missing? I've tried deleting a table and reuploading an updated one with the Supabase website as a test and that seems to update everything, so I'm kinda clueless how to solve this. I'm building the app in python/flask if that gives any clues.  I'm also using the service role api key to interact with my tables since I've enabled RLS.

**Comment by GaryAustin1** [comment]
How does your front end get the data?  If you insert to the table from one method, then the next time your app selects the data from that table it should be reflected unless you have RLS coming into play or some sort of client caching.

**Reply by matthewlyf** [reply]
Sorry I'm still a beginner so I'm not sure if I'll be answering your questions correctly.  So in the backend I'm running a python script to pull the table from supabase (something like this); 
```
def load_jobs_from_db_as_df():
  response = supabase.table('Table_1').select('*').execute()
  response_dict = response.dict()
  jobs = []
  for row in response_dict['data']:
    jobs.append(dict(row))
  df = pd.DataFrame(jobs)
  return(df)
```
After which I take the returned dataframe, extract values based on a specific query and present it on my dynamic webpage (I think webapp might not be the right term). 
```
<div>
  <h4>{{Table_1['col1']}}</h4>
  <div><b>col1:</b> {{Table1['col1']}}</div>
  <div><b>col2:</b> {{Table1['col2']}}</div>
</div>
```
The scripts work perfectly fine for the original csv I've uploaded to supabase but the new inserts are absent despite it appearing in the table.

**Reply by GaryAustin1** [reply]
Not going to be able to help much with your client side stuff as I don't use Python or Colab.
Your Python script reads the database table and you say it is successful for initial data.
Then you insert some more data to the database.
Now you need to do another operation to the database to get the data.  

If you are inserting in the same place you obviously can update in your client code with your inserted data, or .insert().select() will return the data you insert.

But if your insert is happening independently, then you have to tell you client to reload some how, or poll.  Supabase realtime is designed to allow you to monitor tables, but I don't know if it applies to your case or will work in your environment.

**Reply by matthewlyf** [reply]
Thanks, your explanation definitely helps make it a little clearer to me. Just to give more context, my dynamic webpage performs searches within my Supabase table based on a query, and with the new inserts, I'm unable to see the new entry as an option to search for. I'll do some more research about Supabase real time and about reloading or polling(?). Thanks for your insight!

**Reply by GaryAustin1** [reply]
So when a .select() is run you get your data back as it is at the point in time in the database.  Nothing else will happen with that .select() unless you call it again.    If the database changes, the client knows nothing about that change.  The select call you ran before will not run again on its own and no data will be sent to your client (realtime is one option to do that).  

If you run the query you showed again after the insert, it should return all the new data as well.

**Reply by matthewlyf** [reply]
Hmm this actually is giving me an idea of where I can start investigating. I guess my only question is why deleting the table and reuploading an updated table automatically keeps my front end current as opposed to inserting rows.

**Reply by GaryAustin1** [reply]
If you reload the table (call .select()) after your inserts the inserted data should show up.  If you have RLS enabled then you could have issues where RLS is not met, but you have not mentioned RLS or policies.

**Reply by matthewlyf** [reply]
I have RLS set up but no policies, I thought using the service role API key would have let me avoid any of the restrictions when using a table with RLS.

**Reply by GaryAustin1** [reply]
Yeah, if you have service_role from the client then RLS is not involved.

**Reply by matthewlyf** [reply, marked as answer]
I just wanted to say thanks again for attempting to troubleshoot my issue.  I just wanted to update you and admit that I'm actually just dumb AF - turns out the default max rows to be returned by supabase is set to 1000 so I was actually missing a bunch of rows + my insert.  I expanded it which solved everything but I am a bit suspicious about doing that since there is recommendation not to do so.
